File: Backend/src/main/java/org/example/datn_sd69/modules/promotion/service/PromotionService.java
package org.example.datn_sd69.modules.promotion.service;

import org.example.datn_sd69.modules.promotion.dto.request.PromotionRequest;
import org.example.datn_sd69.modules.promotion.dto.response.FlashSaleProductResponse;
import org.example.datn_sd69.modules.promotion.dto.response.PromotionProductVariantOptionResponse;
import org.example.datn_sd69.modules.promotion.dto.response.PromotionResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.time.LocalDateTime;

public interface PromotionService {

    Page<PromotionResponse> getAll(String keyword, Integer status, Pageable pageable);

    PromotionResponse getById(Integer id);

    PromotionResponse create(PromotionRequest request);

    PromotionResponse update(Integer id, PromotionRequest request);

    /** Giữ signature cũ để không làm vỡ caller khác. */
    PromotionResponse changeStatus(Integer id, Integer status);

    /** Bản stale-safe dành cho Admin FE mới. */
    PromotionResponse changeStatus(Integer id, Integer status, String expectedRevision);

    /** Giữ signature cũ để không làm vỡ caller khác. */
    void softDelete(Integer id);

    /** Bản stale-safe dành cho Admin FE mới. */
    void softDelete(Integer id, String expectedRevision);

    int syncExpiredPromotions();

    Page<FlashSaleProductResponse> getActiveFlashSaleProducts(Pageable pageable);

    /**
     * Thời điểm bắt đầu gần nhất của một Flash Sale tương lai.
     *
     * Dùng cho FE đặt timer tự refresh, không dùng để quyết định Promotion active.
     */
    LocalDateTime getNextFlashSaleStartDate();

    Page<PromotionProductVariantOptionResponse> searchProductVariantsForPromotion(
            String keyword,
            LocalDateTime startDate,
            LocalDateTime endDate,
            Integer ignorePromotionId,
            Pageable pageable
    );

}

File: Frontend/src/modules/admin/feature/promotion/services/promotion.service.ts
import api from "@/common/api";
import type {
  FlashSaleProductResponse,
  FlashSaleSearchParams,
  PageResponse,
  ProductVariantSearchParams,
  PromotionProductVariantOptionResponse,
  PromotionRequest,
  PromotionResponse,
  PromotionSearchParams,
  PromotionStatusRequest,
} from "../types/promotion.types";

const normalizeDateTime = (value?: string | null) => {
  if (!value) return undefined;

  if (value.length === 16) {
    return `${value}:00`;
  }

  return value;
};

export const promotionService = {
  getAll(params: PromotionSearchParams = {}) {
    return api.get<PageResponse<PromotionResponse>>("/admin/promotions", {
      params: {
        keyword: params.keyword || undefined,
        status: params.status ?? undefined,
        page: params.page ?? 0,
        size: params.size ?? 10,
      },
    });
  },

  getById(id: number) {
    return api.get<PromotionResponse>(`/admin/promotions/${id}`);
  },

  searchProductVariants(params: ProductVariantSearchParams = {}) {
    return api.get<PageResponse<PromotionProductVariantOptionResponse>>(
      "/admin/promotions/product-variants",
      {
        params: {
          keyword: params.keyword || undefined,
          startDate: normalizeDateTime(params.startDate),
          endDate: normalizeDateTime(params.endDate),
          ignorePromotionId: params.ignorePromotionId ?? undefined,
          page: params.page ?? 0,
          size: params.size ?? 10,
        },
      }
    );
  },

  create(payload: PromotionRequest) {
    return api.post<PromotionResponse>("/admin/promotions", {
      name: payload.name,
      startDate: normalizeDateTime(payload.startDate),
      endDate: normalizeDateTime(payload.endDate),
      variants: payload.variants,
    });
  },

  update(id: number, payload: PromotionRequest) {
    return api.put<PromotionResponse>(`/admin/promotions/${id}`, {
      name: payload.name,
      startDate: normalizeDateTime(payload.startDate),
      endDate: normalizeDateTime(payload.endDate),
      variants: payload.variants,
    });
  },

  changeStatus(id: number, payload: PromotionStatusRequest) {
    return api.patch<PromotionResponse>(
      `/admin/promotions/${id}/status`,
      payload
    );
  },

  remove(id: number) {
    return api.delete(`/admin/promotions/${id}`);
  },

  getFlashSaleProducts(params: FlashSaleSearchParams = {}) {
    return api.get<PageResponse<FlashSaleProductResponse>>(
      "/promotions/flash-sale",
      {
        params: {
          page: params.page ?? 0,
          size: params.size ?? 8,
        },
      }
    );
  },
};